File: 주영/Array/209. Minimum Size Subarray Sum.js
var minSubArrayLen = function (target, nums) {
  let start = 0;
  let sum = 0;
  let length = Number.MAX_SAFE_INTEGER;

  for (let end = 0; end < nums.length; end++) {
    sum += nums[end];
    // console.log('pastSum: ', sum);

    while (sum >= target) {
      length = Math.min(length, end - start + 1);
      sum -= nums[start];
      // console.log('curSum: ', sum);
      start++;
    }
  }

  return length === Number.MAX_SAFE_INTEGER ? 0 : length;
};

let target = 7;
let nums = [2, 3, 1, 2, 4, 3];

minSubArrayLen(target, nums);

File: P/209_Minimum_Size_Subarray_Sum.java
//Two Pointer (Time: O(N) Space: O(1))
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int minLength = 100001;
        int sum = 0;
        
        int i = 0, j = 0;
        while(i <= j && (sum >= target || j < nums.length)) {
            if(sum >= target) {
                sum -= nums[i++];
                minLength = Math.min(minLength, j - i + 1);
                if(minLength == 1)
                    return 1;
            } else {
                sum += nums[j++];
            }
        }
        return minLength == 100001 ? 0 : Math.min(minLength, j - i + 1);
    }
}

//Two Pointer (Time: O(N) Space: O(1))
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int minLength = 100001;
        int i = 0, sum = 0;
        for(int j = 0; j < nums.length; j++) {
            sum += nums[j];
            while(sum >= target) {
                minLength = Math.min(minLength, j - i + 1);
                sum -= nums[i++];
            }
            if(minLength == 1)
                return 1;
        }
        return minLength == 100001 ? 0 : minLength;
    }
}